### ISSUE TITLE
Improve CRL profile specificity in Section 7.2 

### ISSUE THREAD

**Issue by ryancdickson** [cabforum/servercert#419]
#373 and Ballot SC-062 represent significant effort by members of the community to improve the clarity of Section 7.1 ("Certificate Profile") of the BRs. Improved clarity _should_ result in more consistent implementations and less opportunity for mis-issuance resulting from confusion or misunderstanding.

I'd like to understand whether members of the Validation Subcommittee or the broader Server Certificate Working Group think there's value in improving the specificity of Section 7.2 ("CRL Profile") - as certificate consumers become increasingly reliant on Certificate Revocation Lists.

**Comment by CBonnell** [comment]
I definitely see the value of fleshing out the CRL profile more explicitly to ensure that CRL-based revocation information can be processed by user agents as the ecosystem moves away from OCSP.

My opinion is that this work would not involve any sort of validation and is more directly concerned with encoding, so servercert-wg would be the best location to discuss.

**Comment by ryancdickson** [comment]
Draft [proposal](https://github.com/ryancdickson/staging/blob/crl-profile/docs/BR.md#722-crl-and-crl-entry-extensions) for a CRL profile ([diff against profiles ballot](https://github.com/cabforum/servercert/compare/aa9fc5d0b2b59504a31638e880cb81c69aefa018..142623b3e1cea326c868ddb6c2f3a735e6e153de))

@CBonnell - any time for you to share any quick feedback? Given the proposed "[Make OCSP Optional](https://github.com/cabforum/servercert/pull/414)" ballot requires CRLs, I'd like to include a CRL profile to help promote consistency.

How this profile was derived:
- Studied 5280 / x509
- Downloaded the 3,800+ CRLs disclosed to CCADB
- Wrote simple script to asn1 parse the set of CRLs to understand which extensions are used, and how often (counts below).

Known shortcoming: CRLReason code requirements (SC61) not yet integrated. 

| CRL Extensions |	Count |	Percent |
| - | - | - |
|:X509v3 Authority Key Identifier |	3866 |	99.82%|
|:X509v3 CRL Number |	3867 |	99.85%|
|:X509v3 Issuing Distribution Point |	1941 |	50.12%|
|:X509v3 CRL Reason Code |	5185949	| N/A|
|:Invalidity Date |	23 |	0.59%|
|:2.5.29.60 (expiredCertsOnCRL)	| 248 |	6.40%|
|:1.3.6.1.4.1.311.21.4 (CRL_NEXT_PUBLISH) 	| 23 |	0.59%|
|:1.3.6.1.4.1.311.21.14 (CRL_SELF_CDP) |	1 |	0.03%|
|:1.3.6.1.4.1.311.21.1 (MS Certificate Services CA Version) |	22 |	0.57%|
|:X509v3 Issuer Alternative Name	| 13 | 	0.34%|
|:Authority Information Access | 28	| 0.72%|

**Comment by ryancdickson** [comment]
Got some feedback privately, moved the Extensions table up in the presented order.

Non-immutable diff: https://github.com/cabforum/servercert/compare/profiles...ryancdickson:staging:crl-profile

**Comment by CBonnell** [comment]
@ryancdickson, I took a first stab at reviewing and proposing some edits: https://github.com/ryancdickson/staging/pull/1

**Comment by timfromdigicert** [comment]
> I definitely see the value of fleshing out the CRL profile more explicitly to ensure that CRL-based revocation information can be processed by user agents as the ecosystem moves away from OCSP.
> 
> My opinion is that this work would not involve any sort of validation and is more directly concerned with encoding, so servercert-wg would be the best location to discuss.

I agree with this; in fact, as I mentioned in Ottawa, the only way the profiles ballot managed to be in scope for the validation SC charter is because the validation charter mentions certificate contents, and the proposer insisted his work needed to be done in the validation SC.  I actually think CRL profiles are out of scope for the validation SC.

**Comment by timfromdigicert** [comment]
Also, this is somewhat obscure, but it's worth thinking through what "byte for byte identical" means for some of these fields, especially since the revoked certificate might have been misissued (important point!).

Is it really the correct thing to do to copy any such errors into the CRL?  In particular, what do you do with a certificate with a misencoded or badly encoded serial number?  It seems to me that you want the correctly encoded serial number in the CRL, not the misencoded one.

As an example of a serial number encoding issue, consider either failure to include a leading zero when necessary, or excessive leading zeros.

It depends a bit on what common CRL and certificate processing engines will do when presented with these corner cases, which make it a bit tricky to determine the right answer.  You might actually need to include both, I guess (the improperly encoded one and the properly encoded one) to cover all edge cases.   Worth thinking through a bit.

**Comment by barrini** [comment]
Even there´s an issue with the serial number encoding, what should go in the CRL is the serial number (wrongly encoded or not) of the certificate because that´s the unique number associated to that certificate. Including both could make some validation software not clearly process the response correctly.

**Comment by timfromdigicert** [comment]
I agree that including both is probably wrong, and I also think that including the incorrectly encoded version is also probably wrong (because it's wrong ...).

But the point that I'm trying to make here is that that means that the "right" thing to do in at least the rare situation of a misencoded serial number is to put the correctly encoded serial number in the CRL, so the CRL is well-formed.  But this means the serial number in the CRL is *not* byte-for-byte identical to the one in the cert.

**Comment by ryancdickson** [comment]
Thanks for the feedback, all!

Re: Serial # misencodings - I agree with @barrini. 

Making such an exception for misencodings here feels awkward (and it's inconsistent) given we made no such effort in the Certificate Profiles work. Unless there's evidence to suggest misencoded serial #'s are a pervasive problem (and that products consuming CRLs are failing to process those misencoded serial #s), byte-for-byte feels like the most straight forward approach to me.

If such evidence is available, then I'd propose changing to something like: "_MUST be byte-for-byte identical to the serialNumber contained in the revoked certificate, unless the serialNumber contained in the revoked certificate is misencoded. In scenarios where the serialNumber contained in the revoked certificate is misencoded, this value MAY be corrected such that it is not be byte-for-byte identical (i.e., to correct the misencoded value)_."

Included the proposed changes from @CBonnell with some minor formatting changes (thank you @CBonnell!).  

For now, I'm planning on including these changes in the "Make OCSP Optional" Ballot (that I'll look to move into official discussion once SC-62 is merged into Main).  Including the proposed update there makes sense given the proposal requires use of CRLs.

Happy to consider any additional feedback!

**Comment by ryancdickson** [comment]
Closing this issue (profile being proposed in SC-63). Will reopen, if needed (i.e., ballot fails).

**Comment by barrini** [comment]
Thanks

 

De: Ryan Dickson ***@***.***> 
Enviado el: jueves, 27 de abril de 2023 17:51
Para: cabforum/servercert ***@***.***>
CC: Inigo Barreira ***@***.***>; Mention
***@***.***>
Asunto: Re: [cabforum/servercert] Improve CRL profile specificity in Section
7.2 (Issue #419)

 

CAUTION: This email originated from outside of the organization. Do not
click links or open attachments unless you recognize the sender and know the
content is safe.

 

Closing this issue (profile being proposed in SC-63). Will reopen, if needed
(i.e., ballot fails).

-
Reply to this email directly, view it on GitHub
<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.co
m%2Fcabforum%2Fservercert%2Fissues%2F419%23issuecomment-1525938027&data=05%7
C01%7Cinigo.barreira%40sectigo.com%7Cb157e74d4aab4d99644408db473740db%7C0e9c
48946caa465d96604b6968b49fb7%7C0%7C0%7C638182074891269554%7CUnknown%7CTWFpbG
Zsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3
000%7C%7C%7C&sdata=dMC635L3g9fYX8vvM1RZUv4hYcm9C05UXIv0AmNjmMc%3D&reserved=0
> , or unsubscribe
<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.co
m%2Fnotifications%2Funsubscribe-auth%2FAWFQXOJRXXIM7M4ZIONWARLXDKIXXANCNFSM6
AAAAAAUPKPMPQ&data=05%7C01%7Cinigo.barreira%40sectigo.com%7Cb157e74d4aab4d99
644408db473740db%7C0e9c48946caa465d96604b6968b49fb7%7C0%7C0%7C63818207489126
9554%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1
haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=L6O367tnRi5MlGyUAnE%2Fo57hlZQEOigy
HxggzM29c%2BE%3D&reserved=0> .
You are receiving this because you were mentioned.
<https://github.com/notifications/beacon/AWFQXOM5NJYRKAZYOW7DZHLXDKIXXA5CNFS
M6AAAAAAUPKPMPSWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTS
26P3WW.gif> Message ID:
***@***.***
***@***.***> >
